### PR TITLE
Updated IceStorm Parser

### PR DESCRIPTION
**Pull request by InsertCreativityHere**
This PR updates the `icestormadmin` parser to use the latest Bison+Flex, like we did for Slice a couple years ago.
Before this PR it was being generated with Bison `2.7` (released in 2021). Now, it's generated with Bison `3.8`.

It also updates the `Grammar.y` and `Scanner.l` files to use our new style, since `clang-format` is incapable of formatting these files.

Any file named `Scanner.cpp`, `Grammar.cpp`, or `Grammar.h` is generated by Flex/Bison,
and hence probably not worth reviewing.

### CODE REVIEW

File: cpp/src/Slice/Grammar.y [hunk 1]
```diff
@@ -22,6 +22,9 @@
 #define YYMAXDEPTH  10000
 #define YYINITDEPTH YYMAXDEPTH
 
+// Newer bison versions allow to disable stack resizing by defining yyoverflow.
```
Comment (InsertCreativityHere) [inline cpp/src/Slice/Grammar.y:25]: While working on updating `IceStorm`, I saw `yyoverflow` and realized it was missing from the Slice grammar.
We had it on 4.0 (here it is on the 'before_streamline' tag: https://github.com/zeroc-ice/ice/blob/6e08b82a0298696c470137808425893455bc0e76/cpp/src/Slice/Grammar.y#L25-L26).

But I guess it got lost among the backporting, so I added it back. Doing so meant I also had to add another `warning disable` macro, because of how trivial this definition is (it just errors, so some code is now unused)

This function is necessary because we _cannot_ allow flex (the scanner) to reallocate a new stack.
Flex uses simple alloc/memcpy calls which don't work with our C++ types.
Flex supports a C++ mode, but we never switched to it.

File: cpp/src/IceStorm/Parser.h [hunk 2]
```diff
@@ -10,30 +10,6 @@
...
-// assignment operator, etc.
-//
-#define YYMAXDEPTH 10000
-#define YYINITDEPTH YYMAXDEPTH // Initial depth is set to max depth, for the reasons described above.
-
-//
-// Newer bison versions allow to disable stack resizing by defining
-// yyoverflow.
-//
-#define yyoverflow(a, b, c, d, e, f) yyerror(a)
```
Comment (InsertCreativityHere) [inline cpp/src/IceStorm/Parser.h:35]: These definitions were moved into `Grammar.y`, since that's where we actually use them.
This is also where they live for the Slice compiler, so it's more consistent as well.

`Grammar.y` `include`s this file at the top, so moving it shouldn't of affected anything.

File: cpp/src/IceStorm/Parser.h [hunk 3]
```diff
@@ -10,30 +10,6 @@
 #include <list>
 #include <stdio.h>
 
-//
-// Stuff for flex and bison
-//
-
-#define YYSTYPE std::list<std::string>
-#define YY_DECL int yylex(YYSTYPE* yylvalp)
-YY_DECL;
```
Comment (InsertCreativityHere) [inline cpp/src/IceStorm/Parser.h:19]: This code is very outdated. In modern Bison, these lines were replaced by the:
```
%define api.value.type {std::list<std::string>}
```
you'll find was added on line 69 of the the `Grammar.y` file.

File: cpp/src/IceStorm/Grammar.y [hunk 4]
```diff
@@ -46,13 +63,14 @@ yyerror(const char* s)
 
 %}
 
-%pure-parser
+// Directs Bison to generate a re-entrant parser.
+%define api.pure
+// Specifies what type to back the tokens with (their semantic values).
+%define api.value.type {std::list<std::string>}
```
Comment (InsertCreativityHere) [inline cpp/src/IceStorm/Grammar.y:69]: Replaces the outdated way we were specifying token types in `Parser.h`:
```
#define YYSTYPE std::list<std::string>
#define YY_DECL int yylex(YYSTYPE* yylvalp)
YY_DECL;
```

File: cpp/src/IceStorm/Grammar.y [hunk 5]
```diff
@@ -74,7 +92,7 @@ start
 : commands
 {
 }
-|
+| %empty
```
Comment (InsertCreativityHere) [inline cpp/src/IceStorm/Grammar.y:95]: It's better to explicitly indicate empty rules, for readability.

File: cpp/src/IceStorm/Grammar.y [hunk 6]
```diff
@@ -1,37 +1,54 @@
...
+
+%code{
+
+// Forward declaration of the lexing function generated by flex, so bison knows about it.
+// This must match the definition of 'yylex' in the generated scanner.
+int yylex(YYSTYPE* yylvalp);
+
+}
+
+%{
```
Comment (InsertCreativityHere) [inline cpp/src/IceStorm/Grammar.y:32]: This code was moved from `Parser.h`. It's Bison specific code, so it had no business in `Parser.h`.

`requires` just means it's put in a specific spot in the header file, before Bison tries to define these things itself.

File: cpp/src/IceStorm/Grammar.y [hunk 7]
```diff
@@ -1,37 +1,54 @@
-%{
+%code top{
```
Comment (InsertCreativityHere) [inline cpp/src/IceStorm/Grammar.y:1]: We want the copyright to be at the very top of the generated file.

File: cpp/src/IceStorm/Scanner.l [hunk 8]
```diff
@@ -1,64 +1,76 @@
-%{
-
+%top{
```
Comment (InsertCreativityHere) [inline cpp/src/IceStorm/Scanner.l:1]: We want our include and our copyright at the very top of the file.

File: cpp/src/IceStorm/Scanner.l [hunk 9]
```diff
@@ -1,64 +1,76 @@
...
+  /* We always want the scanner to run in interactive mode. */
+%option always-interactive
+
+  /* Disables the generation of functions we don't use to reduce clutter, and possibly improve performance. */
+%option noyywrap
+%option noyy_scan_buffer noyy_scan_bytes noyy_scan_string
+%option noyyget_extra noyyset_extra noyyget_leng noyyget_text
+%option noyyget_in noyyset_in noyyget_out noyyset_out
+%option noyyget_lineno noyyset_lineno noyyget_lloc noyyset_lloc
+%option noyyget_lval noyyset_lval noyyget_debug noyyset_debug
```
Comment (InsertCreativityHere) [inline cpp/src/IceStorm/Scanner.l:68]: Tell Flex not to generate any functions we don't use to cut down on the number of warnings, and make things leaner.

I checked the generated code, and these functions are indeed unused. We do the same thing in:
https://github.com/zeroc-ice/ice/blob/3111cba6b2116e61b479f368f7cc68caf00862e0/cpp/src/Slice/Scanner.l#L92

File: cpp/src/IceStorm/Scanner.l [hunk 10]
```diff
@@ -150,22 +162,22 @@ keyword [[:alpha:]]*
             s += static_cast<char>(c);
         }
     }
-    yylvalp->clear();
```
Comment (InsertCreativityHere) [inline cpp/src/IceStorm/Scanner.l:153]: This is the same variable, it's name was just changed is all.
`yylval` is the correct name for this variable.

File: cpp/include/IceUtil/ScannerConfig.h [hunk 11]
```diff
@@ -7,21 +7,11 @@
 
 #include "Config.h" // Required by generated Scanners.
 
-//
```
Comment (bernardnormier) [inline cpp/include/IceUtil/ScannerConfig.h:10]: The actual fix is to include `<cstdint>` at the top of each Scanner.l. 

I did this in a previous PR for the Slice and IceGrid scanners.

File: cpp/src/IceStorm/Scanner.l (hunk 9 above)
Comment (bernardnormier) [reply cpp/src/IceStorm/Scanner.l:68]: Should we add these options to the IceGrid scanner too?

File: cpp/src/Slice/Grammar.y (hunk 1 above)
Comment (bernardnormier) [reply cpp/src/Slice/Grammar.y:25]: What about making the same update to the IceGrid grammar file?

**Comment by InsertCreativityHere** [comment]
> What about making the same update to the IceGrid grammar file?

Yes, I'm planning on doing that on Monday! Should be straightforward since the IceStorm and IceGrid grammars are very similar. I just wanted to do one parser at a time, and avoid giving you merge conflicts!

**Review by bernardnormier — APPROVED** [review]

File: cpp/src/IceStorm/Scanner.l [hunk 12]
```diff
@@ -1,64 +1,76 @@
...
-// '=' : conversion from 'size_t' to 'int', possible loss of data
-// The result of fread() is a size_t and gets inserted into an int
-//
-#   pragma warning(disable:4267)
-#endif
-
-#if defined(__GNUC__)
-#   pragma GCC diagnostic ignored "-Wsign-compare"
+#ifdef _MSC_VER
+#    ifdef yywrap
```
Comment (pepone) [inline cpp/src/IceStorm/Scanner.l:18]: Why is this required, should we add a comment?

File: cpp/src/IceStorm/Scanner.l [hunk 13]
```diff
@@ -1,64 +1,76 @@
...
 }
-#define         YY_USER_INIT initScanner();
+
+// Override some of the functions flex auto-generates with our own implementations.
+#define YY_USER_INIT initScanner();
+#define YY_INPUT(buf, result, maxSize) parser->getInput(buf, result, maxSize)
 
 %}
 
+  /* Instructs flex to not suppress any warnings when generating the scanner. */
```
Comment (pepone) [inline cpp/src/IceStorm/Scanner.l:43]: Can we use our regular comment style, `//` for non-doc comments? The indentation of the comment looks odd too.

**Review by pepone — APPROVED** [review]

File: cpp/src/IceStorm/Scanner.l [hunk 14]
```diff
@@ -1,64 +1,76 @@
...
-#      define yywrap() 1
-#   endif
-#   define YY_NO_UNISTD_H
-#endif
-
-#define YY_INPUT(buf, result, maxSize) parser->getInput(buf, result, maxSize)
-
 namespace IceStorm
 {
+    typedef map<string, int> StringTokenMap;
```
Comment (externl) [inline cpp/src/IceStorm/Scanner.l:31]:
```suggestion
    using StringTokenMap = map<string, int>;
```

**Review by externl — APPROVED** [review]

File: cpp/src/IceStorm/Scanner.l (hunk 12 above)
Comment (InsertCreativityHere) [reply cpp/src/IceStorm/Scanner.l:18]: I just copy-pasted it because it's always been there.

I'll do some research and see if I can figure out why we have it.

Comment (InsertCreativityHere) [reply cpp/src/IceStorm/Scanner.l]: Seems like we don't actually need this anymore.
Setting the `%noyywrap` option does this same thing but cleaner.

All this is about what the scanner does when it hits EOF.
This function returns `0` to signal it should wait, then load more. Or `1` to signal "this is truly EOF".

File: cpp/src/IceStorm/Scanner.l (hunk 13 above)
Comment (InsertCreativityHere) [reply cpp/src/IceStorm/Scanner.l:35]: Unfortunately not. Flex only supports C-style comments, and this indentation is also necessary.
https://www.cs.virginia.edu/~cr4bd/flex-manual/Comments-in-the-Input.html

File: cpp/src/IceStorm/Scanner.l (hunk 14 above)
Comment (InsertCreativityHere) [reply cpp/src/IceStorm/Scanner.l]: I agree, much better!
